core/src/main/java/com/linecorp/armeria/internal/SslContextUtil.java

```diff
@@ -65,20 +71,45 @@
 
     public static final List<String> DEFAULT_PROTOCOLS = ImmutableList.of("TLSv1.3", "TLSv1.2");
 
+    private static final List<String> DEFAULT_JDKENGINE_CIPHERS = ImmutableList.<String>builder()
+            .addAll(Http2SecurityUtil.CIPHERS)
+            .build();
+
+    private static final List<String> DEFAULT_JDKENGINE_PROTOCOLS = ImmutableList.of("TLSv1.2");
+
     /**
-     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2, TLSv1.3, and
-     * TLSv1.2.
+     * Creates a {@link SslContext} with Armeria's defaults, enabling support for HTTP/2,
+     * TLSv1.3 (if supported), and TLSv1.2.
      */
     public static SslContext createSslContext(Supplier<SslContextBuilder> sslContextSupplier,
                                               boolean forceHttp1,
                                               Consumer<? super SslContextBuilder> userCustomizer) {
         final SslContextBuilder builder = sslContextSupplier.get();
 
-        builder.sslProvider(Flags.useOpenSsl() ? SslProvider.OPENSSL : SslProvider.JDK)
-                  .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]));
-
-        builder.ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
-
+        if (Flags.useOpenSsl()) {
+            builder.sslProvider(SslProvider.OPENSSL)
+                   .protocols(DEFAULT_PROTOCOLS.toArray(new String[0]))
+                   .ciphers(DEFAULT_CIPHERS, SupportedCipherSuiteFilter.INSTANCE);
+        } else {
+            try {
+                // use this to determine actual supported protocols.
+                final SslContextBuilder jdkBuilder = sslContextSupplier.get();
+                final SSLEngine jdkEngine = jdkBuilder.build().newEngine(ByteBufAllocator.DEFAULT);
+                builder.sslProvider(SslProvider.JDK)
```

Comment:
Let's set `protocols` to something like

```java

Set<String> supportedProtocols = ImmutableSet.copyOf(engine.supportedProtocols());
DEFAULT_PROTOCOLS.stream().filter(protocol -> supportedProtocols.contains(protocols)).toArray(String[]:new);
```

Reply:
@anuraaga I am a bit confused. `SslContextUtil` is the builder for for `SSLEngine` & we are using `SSLEngine` to detect feature in `SslContextUtil`. This looks to be cyclic check. May be SSLContextUtil should not set protocol and cipherSuite, thereby relying on SSLEngine defaults. Sorry, if this discussion is going in circle.

Reply:
Address the comment by creating an `SSLEngine` with the default settings to get the support protocol list and filter.